src/pybind/mgr/mgr_module.py

```diff
@@ -812,12 +812,120 @@ def get_localized_module_option(self, key: str, default: OptionValue = None) ->
 ServerInfoT = Dict[str, Union[str, List[ServiceInfoT]]]
 PerfCounterT = Dict[str, Any]
 
+class API:
+    def DecoratorFactory(attr: str, default: Any):
+        class DecoratorClass:
+            _ATTR_TOKEN = f'__ATTR_{attr.upper()}__'
 
-class MgrModule(ceph_module.BaseMgrModule, MgrModuleLoggingMixin):
+            def __init__(self, value=default):
+                self.value = value
+
+            def __call__(self, func):
+                setattr(func, self._ATTR_TOKEN, self.value)
+                return func
+
+            @classmethod
+            def get(cls, func):
+                return getattr(func, cls._ATTR_TOKEN, default)
+
+        return DecoratorClass
+
+    hook = DecoratorFactory('hook', default=False)(True)
+    perm = DecoratorFactory('perm', default='r')
+    internal = DecoratorFactory('internal', default=False)(True)
+
+class MgrModuleHooks:
     COMMANDS = []  # type: List[Any]
-    MODULE_OPTIONS: List[Option] = []
+    MODULE_OPTIONS = []  # type: List[dict]
     MODULE_OPTION_DEFAULTS = {}  # type: Dict[str, Any]
 
+    @API.hook
+    def notify(self, notify_type, notify_id):
```

Comment:
I'd encourage you to add the types from the docstring and put them as type hints (for all the methods moved or modified here):
```suggestion
    def notify(self, notify_type: str, notify_id: str):
```